<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.4.2_09) on Sun Sep 04 19:19:00 CDT 2005 -->
<TITLE>
BooleanAllele (ANJI API)
</TITLE>

<META NAME="keywords" CONTENT="org.jgap.impl.BooleanAllele class">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="BooleanAllele (ANJI API)";
}
</SCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=3 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/BooleanAllele.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV CLASS&nbsp;
&nbsp;<A HREF="../../../org/jgap/impl/ChromosomePool.html" title="class in org.jgap.impl"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../index.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="BooleanAllele.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>

</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
org.jgap.impl</FONT>
<BR>
Class BooleanAllele</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../../resources/inherit.gif" ALT="extended by"><A HREF="../../../org/jgap/Allele.html" title="class in org.jgap">org.jgap.Allele</A>
      <IMG SRC="../../../resources/inherit.gif" ALT="extended by"><B>org.jgap.impl.BooleanAllele</B>
</PRE>
<DL>
<DT><B>All Implemented Interfaces:</B> <DD>java.lang.Comparable</DD>
</DL>
<HR>
<DL>
<DT>public class <B>BooleanAllele</B><DT>extends <A HREF="../../../org/jgap/Allele.html" title="class in org.jgap">Allele</A></DL>

<P>
A Gene implementation that supports two possible values (alleles) for each gene: true and
 false.
 <p>
 NOTE: Since this Gene implementation only supports two different values (true and false),
 there's only a 50% chance that invocation of the setToRandomValue() method will actually
 change the value of this Gene (if it has a value). As a result, it may be desirable to use a
 higher overall mutation rate when this Gene implementation is in use.
<P>

<P>
<HR>

<P>
<!-- ======== NESTED CLASS SUMMARY ======== -->


<!-- =========== FIELD SUMMARY =========== -->

<A NAME="field_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
<B>Field Summary</B></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected static&nbsp;java.lang.Boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/impl/BooleanAllele.html#FALSE_BOOLEAN">FALSE_BOOLEAN</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Shared constant representing the "false" boolean value.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/impl/BooleanAllele.html#m_activeConfiguration">m_activeConfiguration</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The current active configuration that is in use.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;java.lang.Boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/impl/BooleanAllele.html#m_value">m_value</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;References the internal boolean value of this Gene.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected static&nbsp;java.lang.Boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/impl/BooleanAllele.html#TRUE_BOOLEAN">TRUE_BOOLEAN</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Shared constant representing the "true" boolean value.</TD>
</TR>
</TABLE>
&nbsp;
<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../org/jgap/impl/BooleanAllele.html#BooleanAllele()">BooleanAllele</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructs a new BooleanGene with default settings.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../org/jgap/impl/BooleanAllele.html#BooleanAllele(java.lang.Long)">BooleanAllele</A></B>(java.lang.Long&nbsp;an_id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructs a new BooleanGene with default settings.</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
<B>Method Summary</B></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/impl/BooleanAllele.html#booleanValue()">booleanValue</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieves the boolean value of this Gene.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/impl/BooleanAllele.html#cleanup()">cleanup</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Executed by the genetic engine when this Gene instance is no
 
 longer needed and should perform any necessary resource cleanup.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../org/jgap/Allele.html" title="class in org.jgap">Allele</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/impl/BooleanAllele.html#cloneAllele()">cloneAllele</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/impl/BooleanAllele.html#distance(org.jgap.Allele)">distance</A></B>(<A HREF="../../../org/jgap/Allele.html" title="class in org.jgap">Allele</A>&nbsp;target)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/impl/BooleanAllele.html#equals(java.lang.Object)">equals</A></B>(java.lang.Object&nbsp;other)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Compares this BooleanGene with the given object and returns true if
 
 the other object is a BooleanGene and has the same value as this
 
 BooleanGene.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/impl/BooleanAllele.html#getPersistentRepresentation()">getPersistentRepresentation</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieves a string representation of this Gene that includes any
 
 information required to reconstruct it at a later time, such as its
 
 value and internal state.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/impl/BooleanAllele.html#hashCode()">hashCode</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieves the hash code value of this BooleanGene.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/impl/BooleanAllele.html#setToRandomValue(java.util.Random)">setToRandomValue</A></B>(java.util.Random&nbsp;a_numberGenerator)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the value (allele) of this Gene to a random legal value.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/impl/BooleanAllele.html#setValue(java.lang.Object)">setValue</A></B>(java.lang.Object&nbsp;a_newValue)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the value of this Gene to the new given value.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/impl/BooleanAllele.html#setValueFromPersistentRepresentation(java.lang.String)">setValueFromPersistentRepresentation</A></B>(java.lang.String&nbsp;a_representation)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the value and internal state of this Gene from the string
 
 representation returned by a previous invocation of the
 
 getPersistentRepresentation() method.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/impl/BooleanAllele.html#toString()">toString</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieves a string representation of this BooleanGene's value that
 
 may be useful for display purposes.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_org.jgap.Allele"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TD><B>Methods inherited from class org.jgap.<A HREF="../../../org/jgap/Allele.html" title="class in org.jgap">Allele</A></B></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../org/jgap/Allele.html#compareTo(java.lang.Object)">compareTo</A>, <A HREF="../../../org/jgap/Allele.html#getGene()">getGene</A>, <A HREF="../../../org/jgap/Allele.html#getGenes(java.util.Collection)">getGenes</A>, <A HREF="../../../org/jgap/Allele.html#getInnovationId()">getInnovationId</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TD><B>Methods inherited from class java.lang.Object</B></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>clone, finalize, getClass, notify, notifyAll, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->

<A NAME="field_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=1><FONT SIZE="+2">
<B>Field Detail</B></FONT></TD>
</TR>
</TABLE>

<A NAME="TRUE_BOOLEAN"><!-- --></A><H3>
TRUE_BOOLEAN</H3>
<PRE>
protected static final java.lang.Boolean <B>TRUE_BOOLEAN</B></PRE>
<DL>
<DD>Shared constant representing the "true" boolean value. Shared constants
 
 are used to save memory so that a new Boolean object doesn't have to
 
 be constructed each time.
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="FALSE_BOOLEAN"><!-- --></A><H3>
FALSE_BOOLEAN</H3>
<PRE>
protected static final java.lang.Boolean <B>FALSE_BOOLEAN</B></PRE>
<DL>
<DD>Shared constant representing the "false" boolean value. Shared constants
 
 are used to save memory so that a new Boolean object doesn't have to
 
 be constructed each time.
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="m_value"><!-- --></A><H3>
m_value</H3>
<PRE>
protected java.lang.Boolean <B>m_value</B></PRE>
<DL>
<DD>References the internal boolean value of this Gene.
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="m_activeConfiguration"><!-- --></A><H3>
m_activeConfiguration</H3>
<PRE>
protected <A HREF="../../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A> <B>m_activeConfiguration</B></PRE>
<DL>
<DD>The current active configuration that is in use.
<P>
<DL>
</DL>
</DL>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=1><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TD>
</TR>
</TABLE>

<A NAME="BooleanAllele()"><!-- --></A><H3>
BooleanAllele</H3>
<PRE>
public <B>BooleanAllele</B>()</PRE>
<DL>
<DD>Constructs a new BooleanGene with default settings.
<P>
</DL>
<HR>

<A NAME="BooleanAllele(java.lang.Long)"><!-- --></A><H3>
BooleanAllele</H3>
<PRE>
public <B>BooleanAllele</B>(java.lang.Long&nbsp;an_id)</PRE>
<DL>
<DD>Constructs a new BooleanGene with default settings.
<P>
<DT><B>Parameters:</B><DD><CODE>an_id</CODE> - </DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=1><FONT SIZE="+2">
<B>Method Detail</B></FONT></TD>
</TR>
</TABLE>

<A NAME="cloneAllele()"><!-- --></A><H3>
cloneAllele</H3>
<PRE>
public <A HREF="../../../org/jgap/Allele.html" title="class in org.jgap">Allele</A> <B>cloneAllele</B>()</PRE>
<DL>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/Allele.html#cloneAllele()">cloneAllele</A></CODE> in class <CODE><A HREF="../../../org/jgap/Allele.html" title="class in org.jgap">Allele</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>clone of this allele</DL>
</DD>
</DL>
<HR>

<A NAME="setValue(java.lang.Object)"><!-- --></A><H3>
setValue</H3>
<PRE>
public void <B>setValue</B>(java.lang.Object&nbsp;a_newValue)</PRE>
<DL>
<DD>Sets the value of this Gene to the new given value. This class expects the value to be a
 Boolean instance.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_newValue</CODE> - the new value of this Gene instance.</DL>
</DD>
</DL>
<HR>

<A NAME="getPersistentRepresentation()"><!-- --></A><H3>
getPersistentRepresentation</H3>
<PRE>
public java.lang.String <B>getPersistentRepresentation</B>()
                                             throws java.lang.UnsupportedOperationException</PRE>
<DL>
<DD>Retrieves a string representation of this Gene that includes any
 
 information required to reconstruct it at a later time, such as its
 
 value and internal state. This string will be used to represent this
 
 Gene in XML persistence. This is an optional method but, if not
 
 implemented, XML persistence and possibly other features will not be
 
 available. An UnsupportedOperationException should be thrown if no
 
 implementation is provided.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>A string representation of this Gene's current state.
<DT><B>Throws:</B>
<DD><CODE>java.lang.UnsupportedOperationException</CODE> - to indicate that no implementation
 
 is provided for this method.</DL>
</DD>
</DL>
<HR>

<A NAME="setValueFromPersistentRepresentation(java.lang.String)"><!-- --></A><H3>
setValueFromPersistentRepresentation</H3>
<PRE>
public void <B>setValueFromPersistentRepresentation</B>(java.lang.String&nbsp;a_representation)
                                          throws <A HREF="../../../org/jgap/UnsupportedRepresentationException.html" title="class in org.jgap">UnsupportedRepresentationException</A></PRE>
<DL>
<DD>Sets the value and internal state of this Gene from the string
 
 representation returned by a previous invocation of the
 
 getPersistentRepresentation() method. This is an optional method but,
 
 if not implemented, XML persistence and possibly other features will not
 
 be available. An UnsupportedOperationException should be thrown if no
 
 implementation is provided.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_representation</CODE> - the string representation retrieved from a
 
 prior call to the getPersistentRepresentation()
 
 method.
<DT><B>Throws:</B>
<DD><CODE>java.lang.UnsupportedOperationException</CODE> - to indicate that no implementation
 
 is provided for this method.
<DD><CODE><A HREF="../../../org/jgap/UnsupportedRepresentationException.html" title="class in org.jgap">UnsupportedRepresentationException</A></CODE> - if this Gene implementation
 
 does not support the given string representation.</DL>
</DD>
</DL>
<HR>

<A NAME="booleanValue()"><!-- --></A><H3>
booleanValue</H3>
<PRE>
public boolean <B>booleanValue</B>()</PRE>
<DL>
<DD>Retrieves the boolean value of this Gene. This may be more convenient in some cases than the
 more general getAllele() method.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>the boolean value of this Gene.</DL>
</DD>
</DL>
<HR>

<A NAME="setToRandomValue(java.util.Random)"><!-- --></A><H3>
setToRandomValue</H3>
<PRE>
public void <B>setToRandomValue</B>(java.util.Random&nbsp;a_numberGenerator)</PRE>
<DL>
<DD>Sets the value (allele) of this Gene to a random legal value. This
 
 method exists for the benefit of mutation and other operations that
 
 simply desire to randomize the value of a gene.
 
 <p>
 
 NOTE: Since this Gene implementation only supports two different
 
 values (true and false), there's only a 50% chance that invocation
 
 of this method will actually change the value of this Gene (if
 
 it has a value). As a result, it may be desirable to use a higher
 
 overall mutation rate when this Gene implementation is in use.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/Allele.html#setToRandomValue(java.util.Random)">setToRandomValue</A></CODE> in class <CODE><A HREF="../../../org/jgap/Allele.html" title="class in org.jgap">Allele</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_numberGenerator</CODE> - The random number generator that should be
 
 used to create any random values. It's important
 
 to use this generator to maintain the user's
 
 flexibility to configure the genetic engine
 
 to use the random number generator of their
 
 choice.</DL>
</DD>
</DL>
<HR>

<A NAME="equals(java.lang.Object)"><!-- --></A><H3>
equals</H3>
<PRE>
public boolean <B>equals</B>(java.lang.Object&nbsp;other)</PRE>
<DL>
<DD>Compares this BooleanGene with the given object and returns true if
 
 the other object is a BooleanGene and has the same value as this
 
 BooleanGene. Otherwise it returns false.
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE><A HREF="../../../org/jgap/Allele.html#equals(java.lang.Object)">equals</A></CODE> in class <CODE><A HREF="../../../org/jgap/Allele.html" title="class in org.jgap">Allele</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>other</CODE> - the object to compare to this BooleanGene for equality.
<DT><B>Returns:</B><DD>true if this BooleanGene is equal to the given object,
 
 false otherwise.<DT><B>See Also:</B><DD><CODE>Object.equals(java.lang.Object)</CODE></DL>
</DD>
</DL>
<HR>

<A NAME="hashCode()"><!-- --></A><H3>
hashCode</H3>
<PRE>
public int <B>hashCode</B>()</PRE>
<DL>
<DD>Retrieves the hash code value of this BooleanGene.
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE><A HREF="../../../org/jgap/Allele.html#hashCode()">hashCode</A></CODE> in class <CODE><A HREF="../../../org/jgap/Allele.html" title="class in org.jgap">Allele</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>this BooleanGene's hash code.<DT><B>See Also:</B><DD><CODE>Object.hashCode()</CODE></DL>
</DD>
</DL>
<HR>

<A NAME="toString()"><!-- --></A><H3>
toString</H3>
<PRE>
public java.lang.String <B>toString</B>()</PRE>
<DL>
<DD>Retrieves a string representation of this BooleanGene's value that
 
 may be useful for display purposes.
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE><A HREF="../../../org/jgap/Allele.html#toString()">toString</A></CODE> in class <CODE><A HREF="../../../org/jgap/Allele.html" title="class in org.jgap">Allele</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>a string representation of this BooleanGene's value.<DT><B>See Also:</B><DD><CODE>Object.toString()</CODE></DL>
</DD>
</DL>
<HR>

<A NAME="cleanup()"><!-- --></A><H3>
cleanup</H3>
<PRE>
public void <B>cleanup</B>()</PRE>
<DL>
<DD>Executed by the genetic engine when this Gene instance is no
 
 longer needed and should perform any necessary resource cleanup.
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="distance(org.jgap.Allele)"><!-- --></A><H3>
distance</H3>
<PRE>
public double <B>distance</B>(<A HREF="../../../org/jgap/Allele.html" title="class in org.jgap">Allele</A>&nbsp;target)</PRE>
<DL>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/Allele.html#distance(org.jgap.Allele)">distance</A></CODE> in class <CODE><A HREF="../../../org/jgap/Allele.html" title="class in org.jgap">Allele</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>target</CODE> - should be same implementation class and same innovation ID as this gene
<DT><B>Returns:</B><DD>positive distance between genes, where a value closer to 0 represents more similar
 genes; used in computing distance between chromosomes, which in turn is used to compute
 speciation compatibility.<DT><B>See Also:</B><DD><A HREF="../../../org/jgap/Allele.html#distance(org.jgap.Allele)"><CODE>Allele.distance(org.jgap.Allele)</CODE></A></DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=3 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/BooleanAllele.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV CLASS&nbsp;
&nbsp;<A HREF="../../../org/jgap/impl/ChromosomePool.html" title="class in org.jgap.impl"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../index.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="BooleanAllele.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>

</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
