<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.4.2_09) on Sun Sep 04 19:18:58 CDT 2005 -->
<TITLE>
Uses of Class org.jgap.ChromosomeMaterial (ANJI API)
</TITLE>


<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="Uses of Class org.jgap.ChromosomeMaterial (ANJI API)";
}
</SCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=3 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap"><FONT CLASS="NavBarFont1"><B>Class</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../index.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="ChromosomeMaterial.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>

</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<CENTER>
<H2>
<B>Uses of Class<br>org.jgap.ChromosomeMaterial</B></H2>
</CENTER>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Packages that use <A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#com.anji.neat"><B>com.anji.neat</B></A></TD>
<TD><code>com.anji.neat</code> contains classes implementing <a href="http://www.cs.utexas.edu/users/kstanley/neat.html">NEAT</a> framework, including genes, mutation operators, and fitness function used for XOR.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.jgap"><B>org.jgap</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="com.anji.neat"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Uses of <A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A> in <A HREF="../../../com/anji/neat/package-summary.html">com.anji.neat</A></FONT></TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TD COLSPAN=2>Methods in <A HREF="../../../com/anji/neat/package-summary.html">com.anji.neat</A> that return <A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A></CODE></FONT></TD>
<TD><CODE><B>NeatCrossoverReproductionOperator.</B><B><A HREF="../../../com/anji/neat/NeatCrossoverReproductionOperator.html#reproduce(org.jgap.Configuration, org.jgap.Chromosome, org.jgap.Chromosome)">reproduce</A></B>(<A HREF="../../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;config,
          <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;dominantChrom,
          <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;recessiveChrom)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Crossover according to <a
 href="http://nn.cs.utexas.edu/downloads/papers/stanley.ec02.pdf">NEAT </a> crossover
 methodology.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A></CODE></FONT></TD>
<TD><CODE><B>NeatChromosomeUtility.</B><B><A HREF="../../../com/anji/neat/NeatChromosomeUtility.html#newSampleChromosomeMaterial(short, short, short, com.anji.neat.NeatConfiguration, boolean)">newSampleChromosomeMaterial</A></B>(short&nbsp;newNumInputs,
                            short&nbsp;newNumHidden,
                            short&nbsp;newNumOutputs,
                            <A HREF="../../../com/anji/neat/NeatConfiguration.html" title="class in com.anji.neat">NeatConfiguration</A>&nbsp;config,
                            boolean&nbsp;fullyConnected)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;factory method to construct chromosome material for neural net with specified input and
 output dimension, JGAP/NEAT configuration, and amount of connectivity</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TD COLSPAN=2>Methods in <A HREF="../../../com/anji/neat/package-summary.html">com.anji.neat</A> with parameters of type <A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B>WeightMutationOperator.</B><B><A HREF="../../../com/anji/neat/WeightMutationOperator.html#mutate(org.jgap.Configuration, org.jgap.ChromosomeMaterial, java.util.Set, java.util.Set)">mutate</A></B>(<A HREF="../../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;jgapConfig,
       <A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A>&nbsp;target,
       java.util.Set&nbsp;genesToAdd,
       java.util.Set&nbsp;genesToRemove)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Removes from <code>genesToAdd</code> and adds to <code>genesToRemove</code> all
 connection genes that are modified.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SingleTopologicalMutationOperator.</B><B><A HREF="../../../com/anji/neat/SingleTopologicalMutationOperator.html#mutate(org.jgap.Configuration, org.jgap.ChromosomeMaterial, java.util.Set, java.util.Set)">mutate</A></B>(<A HREF="../../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;jgapConfig,
       <A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A>&nbsp;target,
       java.util.Set&nbsp;allelesToAdd,
       java.util.Set&nbsp;allelesToRemove)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B>RemoveConnectionMutationOperator.</B><B><A HREF="../../../com/anji/neat/RemoveConnectionMutationOperator.html#mutate(org.jgap.Configuration, org.jgap.ChromosomeMaterial, java.util.Set, java.util.Set)">mutate</A></B>(<A HREF="../../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;jgapConfig,
       <A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A>&nbsp;target,
       java.util.Set&nbsp;allelesToAdd,
       java.util.Set&nbsp;allelesToRemove)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Removes, in ascending order of weight magnitude, those connections whose weight magnitude is
 less than the maximum weight to be removed.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B>PruneMutationOperator.</B><B><A HREF="../../../com/anji/neat/PruneMutationOperator.html#mutate(org.jgap.Configuration, org.jgap.ChromosomeMaterial, java.util.Set, java.util.Set)">mutate</A></B>(<A HREF="../../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;config,
       <A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A>&nbsp;target,
       java.util.Set&nbsp;genesToAdd,
       java.util.Set&nbsp;genesToRemove)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Traverse network flowing forward and backward to identify unvisited connections and neurons.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B>AddNeuronMutationOperator.</B><B><A HREF="../../../com/anji/neat/AddNeuronMutationOperator.html#mutate(org.jgap.Configuration, org.jgap.ChromosomeMaterial, java.util.Set, java.util.Set)">mutate</A></B>(<A HREF="../../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;jgapConfig,
       <A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A>&nbsp;target,
       java.util.Set&nbsp;allelesToAdd,
       java.util.Set&nbsp;allelesToRemove)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Adds connections according to <a
 href="http://nn.cs.utexas.edu/downloads/papers/stanley.ec02.pdf">NEAT </a> add node mutation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B>AddConnectionMutationOperator.</B><B><A HREF="../../../com/anji/neat/AddConnectionMutationOperator.html#mutate(org.jgap.Configuration, org.jgap.ChromosomeMaterial, java.util.Set, java.util.Set)">mutate</A></B>(<A HREF="../../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;jgapConfig,
       <A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A>&nbsp;target,
       java.util.Set&nbsp;allelesToAdd,
       java.util.Set&nbsp;allelesToRemove)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Adds connections according to <a
 href="http://nn.cs.utexas.edu/downloads/papers/stanley.ec02.pdf">NEAT </a> add connection
 mutation.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.jgap"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Uses of <A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A> in <A HREF="../../../org/jgap/package-summary.html">org.jgap</A></FONT></TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TD COLSPAN=2>Methods in <A HREF="../../../org/jgap/package-summary.html">org.jgap</A> that return <A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected abstract &nbsp;<A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A></CODE></FONT></TD>
<TD><CODE><B>CrossoverReproductionOperator.</B><B><A HREF="../../../org/jgap/CrossoverReproductionOperator.html#reproduce(org.jgap.Configuration, org.jgap.Chromosome, org.jgap.Chromosome)">reproduce</A></B>(<A HREF="../../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;config,
          <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;dominantChrom,
          <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;recessiveChrom)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A></CODE></FONT></TD>
<TD><CODE><B>Configuration.</B><B><A HREF="../../../org/jgap/Configuration.html#getSampleChromosomeMaterial()">getSampleChromosomeMaterial</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieves sample ChromosomeMaterial that contains the desired Gene setup for each
 respective gene position (locus).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A></CODE></FONT></TD>
<TD><CODE><B>ChromosomeMaterial.</B><B><A HREF="../../../org/jgap/ChromosomeMaterial.html#clone(java.lang.Long)">clone</A></B>(java.lang.Long&nbsp;parentId)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns a copy of this ChromosomeMaterial.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A></CODE></FONT></TD>
<TD><CODE><B>ChromosomeMaterial.</B><B><A HREF="../../../org/jgap/ChromosomeMaterial.html#randomInitialChromosomeMaterial(org.jgap.Configuration)">randomInitialChromosomeMaterial</A></B>(<A HREF="../../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_activeConfiguration)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Convenience method that returns a new Chromosome instance with its genes values (alleles)
 randomized.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A></CODE></FONT></TD>
<TD><CODE><B>Chromosome.</B><B><A HREF="../../../org/jgap/Chromosome.html#cloneMaterial()">cloneMaterial</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TD COLSPAN=2>Methods in <A HREF="../../../org/jgap/package-summary.html">org.jgap</A> with parameters of type <A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected abstract &nbsp;void</CODE></FONT></TD>
<TD><CODE><B>MutationOperator.</B><B><A HREF="../../../org/jgap/MutationOperator.html#mutate(org.jgap.Configuration, org.jgap.ChromosomeMaterial, java.util.Set, java.util.Set)">mutate</A></B>(<A HREF="../../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;config,
       <A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A>&nbsp;target,
       java.util.Set&nbsp;allelesToAdd,
       java.util.Set&nbsp;allelesToRemove)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Leaves <code>material</code> unmodified, but updates <code>allelesToAdd</code> and
 <code>allelesToRemove</code> with modifications.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>MutationOperator.</B><B><A HREF="../../../org/jgap/MutationOperator.html#updateMaterial(org.jgap.ChromosomeMaterial, java.util.Set, java.util.Set)">updateMaterial</A></B>(<A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A>&nbsp;material,
               java.util.Set&nbsp;allelesToAdd,
               java.util.Set&nbsp;allelesToRemove)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;updates <code>material</code> with specified sets of alleles; alleles present in both lists
 will be added (or replaced if the gene existed on original material)</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B>Genotype.</B><B><A HREF="../../../org/jgap/Genotype.html#addChromosomeFromMaterial(org.jgap.ChromosomeMaterial)">addChromosomeFromMaterial</A></B>(<A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A>&nbsp;cMat)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>Configuration.</B><B><A HREF="../../../org/jgap/Configuration.html#setSampleChromosomeMaterial(org.jgap.ChromosomeMaterial)">setSampleChromosomeMaterial</A></B>(<A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A>&nbsp;a_sampleChromosomeMaterial)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets sample ChromosomeMaterial that is to be used as a guide for the construction of other
 Chromosomes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B>ChromosomeMaterial.</B><B><A HREF="../../../org/jgap/ChromosomeMaterial.html#distance(org.jgap.ChromosomeMaterial, org.jgap.SpeciationParms)">distance</A></B>(<A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A>&nbsp;target,
         <A HREF="../../../org/jgap/SpeciationParms.html" title="class in org.jgap">SpeciationParms</A>&nbsp;speciationParms)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Calculates compatibility distance between this and <code>target</code> according to <a
 href="http://nn.cs.utexas.edu/downloads/papers/stanley.ec02.pdf">NEAT </a> speciation
 methodology.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TD COLSPAN=2>Constructors in <A HREF="../../../org/jgap/package-summary.html">org.jgap</A> with parameters of type <A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../org/jgap/Chromosome.html#Chromosome(org.jgap.ChromosomeMaterial, java.lang.Long)">Chromosome</A></B>(<A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A>&nbsp;a_material,
           java.lang.Long&nbsp;an_id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this should only be called when a chromosome is being created from persistence; otherwise,
 the ID should be generated by <code>a_activeConfiguration</code>.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=3 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap"><FONT CLASS="NavBarFont1"><B>Class</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../index.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="ChromosomeMaterial.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>

</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
