<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.4.2_09) on Sun Sep 04 19:18:58 CDT 2005 -->
<TITLE>
Uses of Class org.jgap.Chromosome (ANJI API)
</TITLE>


<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="Uses of Class org.jgap.Chromosome (ANJI API)";
}
</SCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=3 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap"><FONT CLASS="NavBarFont1"><B>Class</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../index.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Chromosome.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>

</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<CENTER>
<H2>
<B>Uses of Class<br>org.jgap.Chromosome</B></H2>
</CENTER>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Packages that use <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#com.anji.gomoku"><B>com.anji.gomoku</B></A></TD>
<TD><code>com.anji.gomoku</code> contains classes to handle Go-Moku (see <a href="http://www.gamerz.net/pbmserv/gomoku.html">Gamerz.net</a> for rules).&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#com.anji.imaging"><B>com.anji.imaging</B></A></TD>
<TD><code>com.anji.imaging</code> contains utility classes to handle image processing.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#com.anji.integration"><B>com.anji.integration</B></A></TD>
<TD><code>com.anji.integration</code> contains classes that provide the glue between the various components of the system, including artificial neural net implementations, <a href="http://jgap.sourceforge.net/">JGAP</a>, <a href="http://www.cs.utexas.edu/users/kstanley/neat.html">NEAT</a>, Log4J, and persistence and presentation layers.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#com.anji.neat"><B>com.anji.neat</B></A></TD>
<TD><code>com.anji.neat</code> contains classes implementing <a href="http://www.cs.utexas.edu/users/kstanley/neat.html">NEAT</a> framework, including genes, mutation operators, and fitness function used for XOR.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#com.anji.persistence"><B>com.anji.persistence</B></A></TD>
<TD><code>com.anji.persistence</code> contains classes implementing a simple persistence interface.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#com.anji.polebalance"><B>com.anji.polebalance</B></A></TD>
<TD><code>com.anji.polebalance</code> contains classes to implement pole balancing.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#com.anji.roshambo"><B>com.anji.roshambo</B></A></TD>
<TD><code>com.anji.roshambo</code> contains classes to implement the game of 
<a href="http://en.wikipedia.org/wiki/Roshambo">roshambo</a>, also commonly known as 
rock-scissors-paper.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#com.anji.tournament"><B>com.anji.tournament</B></A></TD>
<TD><code>com.anji.tournament</code> contains classes to handle tournaments and games.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#com.anji.ttt"><B>com.anji.ttt</B></A></TD>
<TD><code>com.anji.ttt</code> contains classes to handle Tic-Tac-Toe (see <a href="http://yahooligans.yahoo.com/content/games/rules/tictactoe/tichistory.html">Yahooligans</a> for rules).&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#examples"><B>examples</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.jgap"><B>org.jgap</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.jgap.impl"><B>org.jgap.impl</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="com.anji.gomoku"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Uses of <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A> in <A HREF="../../../com/anji/gomoku/package-summary.html">com.anji.gomoku</A></FONT></TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TD COLSPAN=2>Methods in <A HREF="../../../com/anji/gomoku/package-summary.html">com.anji.gomoku</A> with parameters of type <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.Object</CODE></FONT></TD>
<TD><CODE><B>GomokuPlayerTranscriber.</B><B><A HREF="../../../com/anji/gomoku/GomokuPlayerTranscriber.html#transcribe(org.jgap.Chromosome)">transcribe</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;c)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../com/anji/tournament/Player.html" title="interface in com.anji.tournament">Player</A></CODE></FONT></TD>
<TD><CODE><B>GomokuPlayerTranscriber.</B><B><A HREF="../../../com/anji/gomoku/GomokuPlayerTranscriber.html#newPlayer(org.jgap.Chromosome)">newPlayer</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;c)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../com/anji/ttt/BoardPlayer.html" title="interface in com.anji.ttt">BoardPlayer</A></CODE></FONT></TD>
<TD><CODE><B>GomokuPlayerTranscriber.</B><B><A HREF="../../../com/anji/gomoku/GomokuPlayerTranscriber.html#newBoardPlayer(org.jgap.Chromosome)">newBoardPlayer</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;c)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="com.anji.imaging"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Uses of <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A> in <A HREF="../../../com/anji/imaging/package-summary.html">com.anji.imaging</A></FONT></TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TD COLSPAN=2>Methods in <A HREF="../../../com/anji/imaging/package-summary.html">com.anji.imaging</A> with parameters of type <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B>IdentifyImageFitnessFunction.</B><B><A HREF="../../../com/anji/imaging/IdentifyImageFitnessFunction.html#evaluate(org.jgap.Chromosome)">evaluate</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;c)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns int between 0 and <code>MAX_FITNESS</code> relative to # images ANN (transcribed
 from <code>c</code>) is able to identify correctly.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="com.anji.integration"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Uses of <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A> in <A HREF="../../../com/anji/integration/package-summary.html">com.anji.integration</A></FONT></TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TD COLSPAN=2>Methods in <A HREF="../../../com/anji/integration/package-summary.html">com.anji.integration</A> with parameters of type <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.Object</CODE></FONT></TD>
<TD><CODE><B>Transcriber.</B><B><A HREF="../../../com/anji/integration/Transcriber.html#transcribe(org.jgap.Chromosome)">transcribe</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;c)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sub-classes must implement this method to convert the genotype to a phenotype.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B>SimpleSelector.</B><B><A HREF="../../../com/anji/integration/SimpleSelector.html#add(org.jgap.Configuration, org.jgap.Chromosome)">add</A></B>(<A HREF="../../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_activeConfigurator,
    <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;a_chromosomeToAdd)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Add <code>a_chromosomeToAdd</code> to set of chromosomes to be evaluated.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.Object</CODE></FONT></TD>
<TD><CODE><B>AnjiNetTranscriber.</B><B><A HREF="../../../com/anji/integration/AnjiNetTranscriber.html#transcribe(org.jgap.Chromosome)">transcribe</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;genotype)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../com/anji/nn/AnjiNet.html" title="class in com.anji.nn">AnjiNet</A></CODE></FONT></TD>
<TD><CODE><B>AnjiNetTranscriber.</B><B><A HREF="../../../com/anji/integration/AnjiNetTranscriber.html#newAnjiNet(org.jgap.Chromosome)">newAnjiNet</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;genotype)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;create new <code>AnjiNet</code> from <code>genotype</code></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../com/anji/integration/Activator.html" title="interface in com.anji.integration">Activator</A></CODE></FONT></TD>
<TD><CODE><B>ActivatorTranscriber.</B><B><A HREF="../../../com/anji/integration/ActivatorTranscriber.html#newActivator(org.jgap.Chromosome)">newActivator</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;ch)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructs <code>Activator</code> phenotype from <code>Chromosome</code> genotype.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.Object</CODE></FONT></TD>
<TD><CODE><B>ActivatorTranscriber.</B><B><A HREF="../../../com/anji/integration/ActivatorTranscriber.html#transcribe(org.jgap.Chromosome)">transcribe</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;c)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TD COLSPAN=2>Constructors in <A HREF="../../../com/anji/integration/package-summary.html">com.anji.integration</A> with parameters of type <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../com/anji/integration/XmlPersistableChromosome.html#XmlPersistableChromosome(org.jgap.Chromosome)">XmlPersistableChromosome</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;aChromosome)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="com.anji.neat"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Uses of <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A> in <A HREF="../../../com/anji/neat/package-summary.html">com.anji.neat</A></FONT></TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TD COLSPAN=2>Methods in <A HREF="../../../com/anji/neat/package-summary.html">com.anji.neat</A> that return <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></CODE></FONT></TD>
<TD><CODE><B>Evolver.</B><B><A HREF="../../../com/anji/neat/Evolver.html#getChamp()">getChamp</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TD COLSPAN=2>Methods in <A HREF="../../../com/anji/neat/package-summary.html">com.anji.neat</A> with parameters of type <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A></CODE></FONT></TD>
<TD><CODE><B>NeatCrossoverReproductionOperator.</B><B><A HREF="../../../com/anji/neat/NeatCrossoverReproductionOperator.html#reproduce(org.jgap.Configuration, org.jgap.Chromosome, org.jgap.Chromosome)">reproduce</A></B>(<A HREF="../../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;config,
          <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;dominantChrom,
          <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;recessiveChrom)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Crossover according to <a
 href="http://nn.cs.utexas.edu/downloads/papers/stanley.ec02.pdf">NEAT </a> crossover
 methodology.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="com.anji.persistence"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Uses of <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A> in <A HREF="../../../com/anji/persistence/package-summary.html">com.anji.persistence</A></FONT></TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TD COLSPAN=2>Methods in <A HREF="../../../com/anji/persistence/package-summary.html">com.anji.persistence</A> that return <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></CODE></FONT></TD>
<TD><CODE><B>Persistence.</B><B><A HREF="../../../com/anji/persistence/Persistence.html#loadChromosome(java.lang.String, org.jgap.Configuration)">loadChromosome</A></B>(java.lang.String&nbsp;id,
               <A HREF="../../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;config)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></CODE></FONT></TD>
<TD><CODE><B>HibernatePersistence.</B><B><A HREF="../../../com/anji/persistence/HibernatePersistence.html#loadChromosome(java.lang.String, org.jgap.Configuration)">loadChromosome</A></B>(java.lang.String&nbsp;id,
               <A HREF="../../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;config)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></CODE></FONT></TD>
<TD><CODE><B>FilePersistence.</B><B><A HREF="../../../com/anji/persistence/FilePersistence.html#chromosomeFromXml(org.jgap.Configuration, java.lang.String)">chromosomeFromXml</A></B>(<A HREF="../../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;config,
                  java.lang.String&nbsp;xml)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></CODE></FONT></TD>
<TD><CODE><B>FilePersistence.</B><B><A HREF="../../../com/anji/persistence/FilePersistence.html#chromosomeFromXml(org.jgap.Configuration, org.w3c.dom.Node)">chromosomeFromXml</A></B>(<A HREF="../../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;config,
                  org.w3c.dom.Node&nbsp;chromNode)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></CODE></FONT></TD>
<TD><CODE><B>FilePersistence.</B><B><A HREF="../../../com/anji/persistence/FilePersistence.html#loadChromosome(java.lang.String, org.jgap.Configuration)">loadChromosome</A></B>(java.lang.String&nbsp;id,
               <A HREF="../../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;config)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TD COLSPAN=2>Methods in <A HREF="../../../com/anji/persistence/package-summary.html">com.anji.persistence</A> with parameters of type <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>Persistence.</B><B><A HREF="../../../com/anji/persistence/Persistence.html#store(org.jgap.Chromosome)">store</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;c)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>HibernatePersistence.</B><B><A HREF="../../../com/anji/persistence/HibernatePersistence.html#store(org.jgap.Chromosome)">store</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;c)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>FilePersistence.</B><B><A HREF="../../../com/anji/persistence/FilePersistence.html#store(org.jgap.Chromosome)">store</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;c)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="com.anji.polebalance"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Uses of <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A> in <A HREF="../../../com/anji/polebalance/package-summary.html">com.anji.polebalance</A></FONT></TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TD COLSPAN=2>Methods in <A HREF="../../../com/anji/polebalance/package-summary.html">com.anji.polebalance</A> with parameters of type <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>DoublePoleBalanceFitnessFunction.</B><B><A HREF="../../../com/anji/polebalance/DoublePoleBalanceFitnessFunction.html#evaluate(org.jgap.Chromosome)">evaluate</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;c)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Evaluate chromosome and set fitness.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="com.anji.roshambo"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Uses of <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A> in <A HREF="../../../com/anji/roshambo/package-summary.html">com.anji.roshambo</A></FONT></TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TD COLSPAN=2>Methods in <A HREF="../../../com/anji/roshambo/package-summary.html">com.anji.roshambo</A> with parameters of type <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../com/anji/roshambo/RoshamboPlayer.html" title="interface in com.anji.roshambo">RoshamboPlayer</A></CODE></FONT></TD>
<TD><CODE><B>RoshamboPlayerTranscriber.</B><B><A HREF="../../../com/anji/roshambo/RoshamboPlayerTranscriber.html#newRoshamboPlayer(org.jgap.Chromosome)">newRoshamboPlayer</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;genotype)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;convenience method to return specific type</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../com/anji/tournament/Player.html" title="interface in com.anji.tournament">Player</A></CODE></FONT></TD>
<TD><CODE><B>RoshamboPlayerTranscriber.</B><B><A HREF="../../../com/anji/roshambo/RoshamboPlayerTranscriber.html#newPlayer(org.jgap.Chromosome)">newPlayer</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;genotype)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.Object</CODE></FONT></TD>
<TD><CODE><B>RoshamboPlayerTranscriber.</B><B><A HREF="../../../com/anji/roshambo/RoshamboPlayerTranscriber.html#transcribe(org.jgap.Chromosome)">transcribe</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;genotype)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="com.anji.tournament"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Uses of <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A> in <A HREF="../../../com/anji/tournament/package-summary.html">com.anji.tournament</A></FONT></TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TD COLSPAN=2>Methods in <A HREF="../../../com/anji/tournament/package-summary.html">com.anji.tournament</A> with parameters of type <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../com/anji/tournament/Player.html" title="interface in com.anji.tournament">Player</A></CODE></FONT></TD>
<TD><CODE><B>PlayerTranscriber.</B><B><A HREF="../../../com/anji/tournament/PlayerTranscriber.html#newPlayer(org.jgap.Chromosome)">newPlayer</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;c)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="com.anji.ttt"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Uses of <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A> in <A HREF="../../../com/anji/ttt/package-summary.html">com.anji.ttt</A></FONT></TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TD COLSPAN=2>Methods in <A HREF="../../../com/anji/ttt/package-summary.html">com.anji.ttt</A> with parameters of type <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.Object</CODE></FONT></TD>
<TD><CODE><B>TttPlayerTranscriber.</B><B><A HREF="../../../com/anji/ttt/TttPlayerTranscriber.html#transcribe(org.jgap.Chromosome)">transcribe</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;c)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../com/anji/tournament/Player.html" title="interface in com.anji.tournament">Player</A></CODE></FONT></TD>
<TD><CODE><B>TttPlayerTranscriber.</B><B><A HREF="../../../com/anji/ttt/TttPlayerTranscriber.html#newPlayer(org.jgap.Chromosome)">newPlayer</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;c)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../com/anji/ttt/BoardPlayer.html" title="interface in com.anji.ttt">BoardPlayer</A></CODE></FONT></TD>
<TD><CODE><B>TttPlayerTranscriber.</B><B><A HREF="../../../com/anji/ttt/TttPlayerTranscriber.html#newBoardPlayer(org.jgap.Chromosome)">newBoardPlayer</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;c)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="examples"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Uses of <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A> in <A HREF="../../../examples/package-summary.html">examples</A></FONT></TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TD COLSPAN=2>Methods in <A HREF="../../../examples/package-summary.html">examples</A> with parameters of type <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B>MinimizingMakeChangeFitnessFunction.</B><B><A HREF="../../../examples/MinimizingMakeChangeFitnessFunction.html#evaluate(org.jgap.Chromosome)">evaluate</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;a_subject)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Determine the fitness of the given Chromosome instance.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B>MinimizingMakeChangeFitnessFunction.</B><B><A HREF="../../../examples/MinimizingMakeChangeFitnessFunction.html#amountOfChange(org.jgap.Chromosome)">amountOfChange</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;a_potentialSolution)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Calculates the total amount of change (in cents) represented by the given potential solution
 and returns that amount.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B>MinimizingMakeChangeFitnessFunction.</B><B><A HREF="../../../examples/MinimizingMakeChangeFitnessFunction.html#getNumberOfCoinsAtGene(org.jgap.Chromosome, int)">getNumberOfCoinsAtGene</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;a_potentialSolution,
                       int&nbsp;a_position)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieves the number of coins represented by the given potential solution at the given gene
 position.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B>MinimizingMakeChangeFitnessFunction.</B><B><A HREF="../../../examples/MinimizingMakeChangeFitnessFunction.html#getTotalNumberOfCoins(org.jgap.Chromosome)">getTotalNumberOfCoins</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;a_potentialsolution)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the total number of coins represented by all of the genes in the given potential
 solution.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.jgap"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Uses of <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A> in <A HREF="../../../org/jgap/package-summary.html">org.jgap</A></FONT></TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TD COLSPAN=2>Methods in <A HREF="../../../org/jgap/package-summary.html">org.jgap</A> that return <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></CODE></FONT></TD>
<TD><CODE><B>Specie.</B><B><A HREF="../../../org/jgap/Specie.html#getRepresentative()">getRepresentative</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></CODE></FONT></TD>
<TD><CODE><B>Specie.</B><B><A HREF="../../../org/jgap/Specie.html#getFittest()">getFittest</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></CODE></FONT></TD>
<TD><CODE><B>Genotype.</B><B><A HREF="../../../org/jgap/Genotype.html#getFittestChromosome()">getFittestChromosome</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieves the Chromosome in the population with the highest fitness value.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TD COLSPAN=2>Methods in <A HREF="../../../org/jgap/package-summary.html">org.jgap</A> with parameters of type <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>Specie.</B><B><A HREF="../../../org/jgap/Specie.html#add(org.jgap.Chromosome)">add</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;aChromosome)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B>Specie.</B><B><A HREF="../../../org/jgap/Specie.html#getChromosomeFitnessValue(org.jgap.Chromosome)">getChromosomeFitnessValue</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;aChromosome)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B>Specie.</B><B><A HREF="../../../org/jgap/Specie.html#match(org.jgap.Chromosome)">match</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;aChromosome)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected abstract &nbsp;void</CODE></FONT></TD>
<TD><CODE><B>NaturalSelector.</B><B><A HREF="../../../org/jgap/NaturalSelector.html#add(org.jgap.Configuration, org.jgap.Chromosome)">add</A></B>(<A HREF="../../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;config,
    <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;c)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B>Genotype.</B><B><A HREF="../../../org/jgap/Genotype.html#addChromosome(org.jgap.Chromosome)">addChromosome</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;chrom)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;add chromosome to population and to appropriate specie</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B>FitnessFunction.</B><B><A HREF="../../../org/jgap/FitnessFunction.html#getFitnessValue(org.jgap.Chromosome)">getFitnessValue</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;a_subject)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Retrieves the fitness value of the given Chromosome.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected abstract &nbsp;int</CODE></FONT></TD>
<TD><CODE><B>FitnessFunction.</B><B><A HREF="../../../org/jgap/FitnessFunction.html#evaluate(org.jgap.Chromosome)">evaluate</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;a_subject)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;Determine the fitness of the given Chromosome instance.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected abstract &nbsp;<A HREF="../../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A></CODE></FONT></TD>
<TD><CODE><B>CrossoverReproductionOperator.</B><B><A HREF="../../../org/jgap/CrossoverReproductionOperator.html#reproduce(org.jgap.Configuration, org.jgap.Chromosome, org.jgap.Chromosome)">reproduce</A></B>(<A HREF="../../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;config,
          <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;dominantChrom,
          <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;recessiveChrom)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B>Chromosome.</B><B><A HREF="../../../org/jgap/Chromosome.html#distance(org.jgap.Chromosome, org.jgap.SpeciationParms)">distance</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;target,
         <A HREF="../../../org/jgap/SpeciationParms.html" title="class in org.jgap">SpeciationParms</A>&nbsp;parms)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Calculates compatibility distance between this and <code>target</code> according to <a
 href="http://nn.cs.utexas.edu/downloads/papers/stanley.ec02.pdf">NEAT </a> speciation
 methodology.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TD COLSPAN=2>Constructors in <A HREF="../../../org/jgap/package-summary.html">org.jgap</A> with parameters of type <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../org/jgap/Specie.html#Specie(org.jgap.SpeciationParms, org.jgap.Chromosome)">Specie</A></B>(<A HREF="../../../org/jgap/SpeciationParms.html" title="class in org.jgap">SpeciationParms</A>&nbsp;aSpeciationParms,
       <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;aRepresentative)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Create new specie from representative.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.jgap.impl"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Uses of <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A> in <A HREF="../../../org/jgap/impl/package-summary.html">org.jgap.impl</A></FONT></TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TD COLSPAN=2>Methods in <A HREF="../../../org/jgap/impl/package-summary.html">org.jgap.impl</A> that return <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></CODE></FONT></TD>
<TD><CODE><B>ChromosomePool.</B><B><A HREF="../../../org/jgap/impl/ChromosomePool.html#acquireChromosome()">acquireChromosome</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Attempts to acquire an Chromosome instance from the chromosome pool.</TD>
</TR>
</TABLE>
&nbsp;
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableSubHeadingColor">
<TD COLSPAN=2>Methods in <A HREF="../../../org/jgap/impl/package-summary.html">org.jgap.impl</A> with parameters of type <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B>WeightedRouletteSelector.</B><B><A HREF="../../../org/jgap/impl/WeightedRouletteSelector.html#add(org.jgap.Configuration, org.jgap.Chromosome)">add</A></B>(<A HREF="../../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_activeConfigurator,
    <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;a_chromosomeToAdd)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Add a Chromosome instance to this selector's working pool of Chromosomes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B>ChromosomePool.</B><B><A HREF="../../../org/jgap/impl/ChromosomePool.html#releaseChromosome(org.jgap.Chromosome)">releaseChromosome</A></B>(<A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;a_chromosome)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Releases a Chromosome to the pool.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=3 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../org/jgap/Chromosome.html" title="class in org.jgap"><FONT CLASS="NavBarFont1"><B>Class</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../index.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Chromosome.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>

</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
