<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN""http://www.w3.org/TR/REC-html40/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc on Fri May 24 15:31:06 CDT 2002 -->
<TITLE>
Chromosome (JGAP Java documentation)
</TITLE>
<LINK REL ="stylesheet" TYPE="text/css" HREF="../../stylesheet.css" TITLE="Style">
</HEAD>
<SCRIPT>
function asd()
{
parent.document.title="Chromosome (JGAP Java documentation)";
}
</SCRIPT>
<BODY BGCOLOR="white" onload="asd();">

<!-- ========== START OF NAVBAR ========== -->
<A NAME="navbar_top"><!-- --></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0">
<TR>
<TD COLSPAN=3 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV CLASS&nbsp;
&nbsp;<A HREF="../../org/jgap/FitnessFunction.html"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html" TARGET="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Chromosome.html" TARGET="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;
<SCRIPT>
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html" TARGET=""><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
<A HREF="../../allclasses-noframe.html" TARGET=""><B>All Classes</B></A>
</NOSCRIPT>
</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<!-- =========== END OF NAVBAR =========== -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
org.jgap</FONT>
<BR>
Class  Chromosome</H2>
<PRE>
java.lang.Object
  |
  +--<B>org.jgap.Chromosome</B>
</PRE>
<HR>
<DL>
<DT>public class <B>Chromosome</B><DT>extends java.lang.Object</DL>

<P>
Chromosomes represent fixed-length collections of genes.
 In the current implementation, this is the lowest
 level class since genes are, themselves, represented
 as individual bits. Each gene can have two values
 (true or false). This should be suitable for most
 applications, but it's always possible for a fitness
 function to consider multiple genes as a single unit
 if more than two possible values are desired.
 In the future, a Gene class and Allele interface may be
 created that will allow more flexibility in this regard.
 <p>
 Chromosomes support reproduction, crossover, and mutation.
 Additional mechanisms for chromosome evolution are
 planned for future versions.
<P>

<P>
<DL>
<DT><B>Author:</B><DD>Neil Rotstan (neil at bluesock.org)</DD>
</DD>
</DL>
<HR>

<P>
<!-- ======== NESTED CLASS SUMMARY ======== -->


<!-- =========== FIELD SUMMARY =========== -->


<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" CELLPADDING="3" CELLSPACING="0" WIDTH="100%">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#Chromosome(java.util.BitSet, int)">Chromosome</A></B>(java.util.BitSet&nbsp;initialGenes,
           int&nbsp;chromosomeLength)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructs this Chromosome instance with the given set of genes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#Chromosome(java.util.BitSet, int, int)">Chromosome</A></B>(java.util.BitSet&nbsp;initialGenes,
           int&nbsp;chromosomeLength,
           int&nbsp;desiredMutationRate)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructs this Chromosome instance with the given set of genes
 and the given mutation rate.</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" CELLPADDING="3" CELLSPACING="0" WIDTH="100%">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
<B>Method Summary</B></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#crossover(org.jgap.Chromosome)">crossover</A></B>(org.jgap.Chromosome&nbsp;mate)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Performs basic crossover between this Chromosome instance and
 the given instance.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#getAllele(int)">getAllele</A></B>(int&nbsp;locus)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the allele (value) of the gene at the given locus
 (index) within the Chromosome.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#mutate()">mutate</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Runs through the genes of this Chromosome, possibly mutating
 some in the process.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;org.jgap.Chromosome</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#randomInitialChromosome(int)">randomInitialChromosome</A></B>(int&nbsp;size)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Convenience method that returns a newly constructed Chromosome
 instance of the given size with a random population of genes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;org.jgap.Chromosome</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#randomInitialChromosome(int, int)">randomInitialChromosome</A></B>(int&nbsp;size,
                        int&nbsp;mutationRate)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Convenience method that returns a newly constructed Chromosome
 instance of the given size with a random population of genes
 and instantiated with the given mutation rate.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;org.jgap.Chromosome</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#reproduce()">reproduce</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns a copy of this Chromosome.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#size()">size</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the size of this Chromosome (the number of genes).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#toString()">toString</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns a string representation of this Chromosome, useful
 for debugging purposes.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" CELLPADDING="3" CELLSPACING="0" WIDTH="100%">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TD><B>Methods inherited from class java.lang.Object</B></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>equals, getClass, hashCode, notify, notifyAll, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->


<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" CELLPADDING="3" CELLSPACING="0" WIDTH="100%">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=1><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TD>
</TR>
</TABLE>

<A NAME="Chromosome(java.util.BitSet, int)"><!-- --></A><H3>
Chromosome</H3>
<PRE>
public <B>Chromosome</B>(java.util.BitSet&nbsp;initialGenes,
                  int&nbsp;chromosomeLength)</PRE>
<DL>
<DD>Constructs this Chromosome instance with the given set of genes.
 Each gene is represented by a single bit in the BitSet. For
 convenience, the randomInitialChromosome method is provided
 that will take care of generating a Chromosome instance of a
 specified size. This constructor exists in case it's desirable
 to initialize Chromosomes with specific gene values.
<P>
<DT><B>Parameters:</B><DD><CODE>initialGenes</CODE> - The set of genes with which to initialize
                      this Chromosome instance. Each bit in the
                      BitSet represents a single gene.</DL>
<HR>

<A NAME="Chromosome(java.util.BitSet, int, int)"><!-- --></A><H3>
Chromosome</H3>
<PRE>
public <B>Chromosome</B>(java.util.BitSet&nbsp;initialGenes,
                  int&nbsp;chromosomeLength,
                  int&nbsp;desiredMutationRate)</PRE>
<DL>
<DD>Constructs this Chromosome instance with the given set of genes
 and the given mutation rate. Each gene is represented by a
 single bit in the BitSet. For convenience, 
 the randomInitialChromosome method is provided that will take
 care of generating a Chromosome instance of a specified size.
 This constructor exists in case it's desirable to initialize
 Chromosomes with specific gene values.
<P>
<DT><B>Parameters:</B><DD><CODE>initialGenes</CODE> - The set of genes with which to initialize
                      this Chromosome instance. Each bit in the
                      BitSet represents a single gene.<DD><CODE>desiredMutationRate</CODE> - The desired chances of mutation
                             expressed as the fraction 1/n where
                             n is this parameter. For example, if
                             1000 were given, then statistically
                             one out of every thousand genes 
                             processed would incur mutation during
                             evolution.</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" CELLPADDING="3" CELLSPACING="0" WIDTH="100%">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=1><FONT SIZE="+2">
<B>Method Detail</B></FONT></TD>
</TR>
</TABLE>

<A NAME="reproduce()"><!-- --></A><H3>
reproduce</H3>
<PRE>
public org.jgap.Chromosome <B>reproduce</B>()</PRE>
<DL>
<DD>Returns a copy of this Chromosome. The returned instance can
 evolve independently of this instance.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>A copy of this Chromosome.</DL>
</DD>
</DL>
<HR>

<A NAME="crossover(org.jgap.Chromosome)"><!-- --></A><H3>
crossover</H3>
<PRE>
public void <B>crossover</B>(org.jgap.Chromosome&nbsp;mate)</PRE>
<DL>
<DD>Performs basic crossover between this Chromosome instance and
 the given instance. A locus (element index) is randomly selected
 and all genes with a locus greater or equal to that randomly
 selected are then swapped between the two Chromosome instances.
 For example, suppose this Chromosome had the genes 1010 and the
 given Chromosome had the genes 0011. This method would choose a
 random index from 0-3. Let's say it chose 2. Then it would swap
 the genes with indexes >= 2 between the two chromosomes. This
 chromosome would then be 1011 and the given Chromosome would be
 0010. 

 Note that the given Chromosome must have the same number of
 genes as this Chromosome or an IllegalArgumentException will be
 thrown.
<P>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.lang.IllegalArgumentException</CODE> - if the size of the given
         Chromosome is not the same as the size of this
         Chromosome.</DL>
</DD>
</DL>
<HR>

<A NAME="mutate()"><!-- --></A><H3>
mutate</H3>
<PRE>
public void <B>mutate</B>()</PRE>
<DL>
<DD>Runs through the genes of this Chromosome, possibly mutating
 some in the process. For each gene, a random number is chosen
 between zero and the mutation rate provided at construction
 (or the default rate if none was provided). If the value of
 the random number is 0, then the gene's bit will be flipped.
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getAllele(int)"><!-- --></A><H3>
getAllele</H3>
<PRE>
public boolean <B>getAllele</B>(int&nbsp;locus)</PRE>
<DL>
<DD>Returns the allele (value) of the gene at the given locus
 (index) within the Chromosome. The first gene is at locus
 zero and the last gene is at the locus equal to the size
 of this Chromosome - 1.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The boolean value of the indicated gene.</DL>
</DD>
</DL>
<HR>

<A NAME="size()"><!-- --></A><H3>
size</H3>
<PRE>
public int <B>size</B>()</PRE>
<DL>
<DD>Returns the size of this Chromosome (the number of genes).
 A Chromosome's size is constant and will never change.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>The number of genes in this Chromosome instance.</DL>
</DD>
</DL>
<HR>

<A NAME="toString()"><!-- --></A><H3>
toString</H3>
<PRE>
public java.lang.String <B>toString</B>()</PRE>
<DL>
<DD>Returns a string representation of this Chromosome, useful
 for debugging purposes.
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE>toString</CODE> in class <CODE>java.lang.Object</CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>A string representation of this Chromosome.</DL>
</DD>
</DL>
<HR>

<A NAME="randomInitialChromosome(int)"><!-- --></A><H3>
randomInitialChromosome</H3>
<PRE>
public static org.jgap.Chromosome <B>randomInitialChromosome</B>(int&nbsp;size)</PRE>
<DL>
<DD>Convenience method that returns a newly constructed Chromosome
 instance of the given size with a random population of genes.
 This method will instantiate the Chromosome with the default
 mutation rate.
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="randomInitialChromosome(int, int)"><!-- --></A><H3>
randomInitialChromosome</H3>
<PRE>
public static org.jgap.Chromosome <B>randomInitialChromosome</B>(int&nbsp;size,
                                                          int&nbsp;mutationRate)</PRE>
<DL>
<DD>Convenience method that returns a newly constructed Chromosome
 instance of the given size with a random population of genes
 and instantiated with the given mutation rate.
<P>
<DD><DL>
</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>

<!-- ========== START OF NAVBAR ========== -->
<A NAME="navbar_bottom"><!-- --></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0">
<TR>
<TD COLSPAN=3 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV CLASS&nbsp;
&nbsp;<A HREF="../../org/jgap/FitnessFunction.html"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html" TARGET="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Chromosome.html" TARGET="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;
<SCRIPT>
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html" TARGET=""><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
<A HREF="../../allclasses-noframe.html" TARGET=""><B>All Classes</B></A>
</NOSCRIPT>
</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<!-- =========== END OF NAVBAR =========== -->

<HR>

</BODY>
</HTML>
