<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.4.2_09) on Sun Sep 04 19:19:00 CDT 2005 -->
<TITLE>
Genotype (ANJI API)
</TITLE>

<META NAME="keywords" CONTENT="org.jgap.Genotype class">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="Genotype (ANJI API)";
}
</SCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=3 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/Genotype.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../org/jgap/Gene.html" title="class in org.jgap"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../org/jgap/IdFactory.html" title="class in org.jgap"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Genotype.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>

</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
org.jgap</FONT>
<BR>
Class Genotype</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../resources/inherit.gif" ALT="extended by"><B>org.jgap.Genotype</B>
</PRE>
<DL>
<DT><B>All Implemented Interfaces:</B> <DD>java.io.Serializable</DD>
</DL>
<HR>
<DL>
<DT>public class <B>Genotype</B><DT>extends java.lang.Object<DT>implements java.io.Serializable</DL>

<P>
Genotypes are fixed-length populations of chromosomes. As an instance of a
 <code>Genotype</code> is evolved, all of its <code>Chromosome</code> objects are also
 evolved. A <code>Genotype</code> may be constructed normally, whereby an array of
 <code>Chromosome</code> objects must be provided, or the static
 <code>randomInitialGenotype()</code> method can be used to generate a <code>Genotype</code>
 with a randomized <code>Chromosome</code> population. Changes made by Tucker and James for
 <a href="http://anji.sourceforge.net/">ANJI </a>:
 <ul>
 <li>added species</li>
 <li>modified order of operations in <code>evolve()</code></li>
 <li>added <code>addChromosome*()</code> methods</li>
 </ul>
<P>

<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../serialized-form.html#org.jgap.Genotype">Serialized Form</A></DL>
<HR>

<P>
<!-- ======== NESTED CLASS SUMMARY ======== -->


<!-- =========== FIELD SUMMARY =========== -->

<A NAME="field_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
<B>Field Summary</B></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Genotype.html#m_activeConfiguration">m_activeConfiguration</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The current active Configuration instance.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;java.util.List</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Genotype.html#m_chromosomes">m_chromosomes</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Chromosomes that makeup thie Genotype's population.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;java.util.List</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Genotype.html#m_species">m_species</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Species that makeup this Genotype's population.</TD>
</TR>
</TABLE>
&nbsp;
<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../org/jgap/Genotype.html#Genotype(org.jgap.Configuration, java.util.List)">Genotype</A></B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_activeConfiguration,
         java.util.List&nbsp;a_initialChromosomes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This constructor is used for random initial Genotypes.</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
<B>Method Summary</B></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Genotype.html#addChromosome(org.jgap.Chromosome)">addChromosome</A></B>(<A HREF="../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;chrom)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;add chromosome to population and to appropriate specie</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Genotype.html#addChromosomeFromMaterial(org.jgap.ChromosomeMaterial)">addChromosomeFromMaterial</A></B>(<A HREF="../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A>&nbsp;cMat)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Genotype.html#addChromosomes(java.util.Collection)">addChromosomes</A></B>(java.util.Collection&nbsp;chromosomes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Genotype.html#addChromosomesFromMaterial(java.util.Collection)">addChromosomesFromMaterial</A></B>(java.util.Collection&nbsp;chromosomeMaterial)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Genotype.html#equals(java.lang.Object)">equals</A></B>(java.lang.Object&nbsp;other)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Compares this Genotype against the specified object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Genotype.html#evolve()">evolve</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Performs one generation cycle, evaluating fitness, selecting survivors, repopulting with
 offspring, and mutating new population.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.List</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Genotype.html#getChromosomes()">getChromosomes</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Genotype.html#getFittestChromosome()">getFittestChromosome</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieves the Chromosome in the population with the highest fitness value.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.List</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Genotype.html#getSpecies()">getSpecies</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../org/jgap/Genotype.html" title="class in org.jgap">Genotype</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Genotype.html#randomInitialGenotype(org.jgap.Configuration)">randomInitialGenotype</A></B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_activeConfiguration)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Convenience method that returns a newly constructed Genotype instance configured according to
 the given Configuration instance.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Genotype.html#toString()">toString</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TD><B>Methods inherited from class java.lang.Object</B></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>clone, finalize, getClass, hashCode, notify, notifyAll, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->

<A NAME="field_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=1><FONT SIZE="+2">
<B>Field Detail</B></FONT></TD>
</TR>
</TABLE>

<A NAME="m_activeConfiguration"><!-- --></A><H3>
m_activeConfiguration</H3>
<PRE>
protected <A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A> <B>m_activeConfiguration</B></PRE>
<DL>
<DD>The current active Configuration instance.
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="m_species"><!-- --></A><H3>
m_species</H3>
<PRE>
protected java.util.List <B>m_species</B></PRE>
<DL>
<DD>Species that makeup this Genotype's population.
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="m_chromosomes"><!-- --></A><H3>
m_chromosomes</H3>
<PRE>
protected java.util.List <B>m_chromosomes</B></PRE>
<DL>
<DD>Chromosomes that makeup thie Genotype's population.
<P>
<DL>
</DL>
</DL>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=1><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TD>
</TR>
</TABLE>

<A NAME="Genotype(org.jgap.Configuration, java.util.List)"><!-- --></A><H3>
Genotype</H3>
<PRE>
public <B>Genotype</B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_activeConfiguration,
                java.util.List&nbsp;a_initialChromosomes)
         throws <A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></PRE>
<DL>
<DD>This constructor is used for random initial Genotypes. Note that the Configuration object
 must be in a valid state when this method is invoked, or a InvalidconfigurationException will
 be thrown.
<P>
<DT><B>Parameters:</B><DD><CODE>a_activeConfiguration</CODE> - The current active Configuration object.<DD><CODE>a_initialChromosomes</CODE> - <code>List</code> contains Chromosome objects: The Chromosome
 population to be managed by this Genotype instance.
<DT><B>Throws:</B>
<DD><CODE>java.lang.IllegalArgumentException</CODE> - if either the given Configuration object or the array of
 Chromosomes is null, or if any of the Genes in the array of Chromosomes is null.
<DD><CODE><A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></CODE> - if the given Configuration object is in an invalid
 state.</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=1><FONT SIZE="+2">
<B>Method Detail</B></FONT></TD>
</TR>
</TABLE>

<A NAME="addChromosomes(java.util.Collection)"><!-- --></A><H3>
addChromosomes</H3>
<PRE>
protected void <B>addChromosomes</B>(java.util.Collection&nbsp;chromosomes)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>chromosomes</CODE> - <code>Collection</code> contains Chromosome objects<DT><B>See Also:</B><DD><A HREF="../../org/jgap/Genotype.html#addChromosome(org.jgap.Chromosome)"><CODE>addChromosome(Chromosome)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="addChromosomesFromMaterial(java.util.Collection)"><!-- --></A><H3>
addChromosomesFromMaterial</H3>
<PRE>
protected void <B>addChromosomesFromMaterial</B>(java.util.Collection&nbsp;chromosomeMaterial)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>chromosomeMaterial</CODE> - <code>Collection</code> contains ChromosomeMaterial objects<DT><B>See Also:</B><DD><A HREF="../../org/jgap/Genotype.html#addChromosomeFromMaterial(org.jgap.ChromosomeMaterial)"><CODE>addChromosomeFromMaterial(ChromosomeMaterial)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="addChromosomeFromMaterial(org.jgap.ChromosomeMaterial)"><!-- --></A><H3>
addChromosomeFromMaterial</H3>
<PRE>
protected void <B>addChromosomeFromMaterial</B>(<A HREF="../../org/jgap/ChromosomeMaterial.html" title="class in org.jgap">ChromosomeMaterial</A>&nbsp;cMat)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>cMat</CODE> - chromosome material from which to construct new chromosome object<DT><B>See Also:</B><DD><A HREF="../../org/jgap/Genotype.html#addChromosome(org.jgap.Chromosome)"><CODE>addChromosome(Chromosome)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="addChromosome(org.jgap.Chromosome)"><!-- --></A><H3>
addChromosome</H3>
<PRE>
protected void <B>addChromosome</B>(<A HREF="../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A>&nbsp;chrom)</PRE>
<DL>
<DD>add chromosome to population and to appropriate specie
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>chrom</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="getChromosomes()"><!-- --></A><H3>
getChromosomes</H3>
<PRE>
public java.util.List <B>getChromosomes</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>List contains Chromosome objects, the population of Chromosomes.</DL>
</DD>
</DL>
<HR>

<A NAME="getSpecies()"><!-- --></A><H3>
getSpecies</H3>
<PRE>
public java.util.List <B>getSpecies</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>List contains Specie objects</DL>
</DD>
</DL>
<HR>

<A NAME="getFittestChromosome()"><!-- --></A><H3>
getFittestChromosome</H3>
<PRE>
public <A HREF="../../org/jgap/Chromosome.html" title="class in org.jgap">Chromosome</A> <B>getFittestChromosome</B>()</PRE>
<DL>
<DD>Retrieves the Chromosome in the population with the highest fitness value.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>The Chromosome with the highest fitness value, or null if there are no chromosomes in
 this Genotype.</DL>
</DD>
</DL>
<HR>

<A NAME="evolve()"><!-- --></A><H3>
evolve</H3>
<PRE>
public void <B>evolve</B>()</PRE>
<DL>
<DD>Performs one generation cycle, evaluating fitness, selecting survivors, repopulting with
 offspring, and mutating new population. This is a modified version of original JGAP method
 which changes order of operations and splits <code>GeneticOperator</code> into
 <code>ReproductionOperator</code> and <code>MutationOperator</code>. New order of
 operations:
 <ol>
 <li>assign <b>fitness </b> to all members of population with
 <code>BulkFitnessFunction</code> or <code>FitnessFunction</code></li>
 <li><b>select </b> survivors and remove casualties from population</li>
 <li>re-fill population with offspring via <b>reproduction </b> operators</li>
 <li><b>mutate </b> offspring (note, survivors are passed on un-mutated)</li>
 </ol>
 Genetic event <code>GeneticEvent.GENOTYPE_EVALUATED_EVENT</code> is fired between steps 2
 and 3. Genetic event <code>GeneticEvent.GENOTYPE_EVOLVED_EVENT</code> is fired after step
 4.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="toString()"><!-- --></A><H3>
toString</H3>
<PRE>
public java.lang.String <B>toString</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD><code>String</code> representation of this <code>Genotype</code> instance.</DL>
</DD>
</DL>
<HR>

<A NAME="randomInitialGenotype(org.jgap.Configuration)"><!-- --></A><H3>
randomInitialGenotype</H3>
<PRE>
public static <A HREF="../../org/jgap/Genotype.html" title="class in org.jgap">Genotype</A> <B>randomInitialGenotype</B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_activeConfiguration)
                                      throws <A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></PRE>
<DL>
<DD>Convenience method that returns a newly constructed Genotype instance configured according to
 the given Configuration instance. The population of Chromosomes will created according to the
 setup of the sample Chromosome in the Configuration object, but the gene values (alleles)
 will be set to random legal values.
 <p>
 Note that the given Configuration instance must be in a valid state at the time this method
 is invoked, or an InvalidConfigurationException will be thrown.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_activeConfiguration</CODE> - 
<DT><B>Returns:</B><DD>A newly constructed Genotype instance.
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></CODE> - if the given Configuration instance not in a valid
 state.</DL>
</DD>
</DL>
<HR>

<A NAME="equals(java.lang.Object)"><!-- --></A><H3>
equals</H3>
<PRE>
public boolean <B>equals</B>(java.lang.Object&nbsp;other)</PRE>
<DL>
<DD>Compares this Genotype against the specified object. The result is true if the argument is an
 instance of the Genotype class, has exactly the same number of chromosomes as the given
 Genotype, and, for each Chromosome in this Genotype, there is an equal chromosome in the
 given Genotype. The chromosomes do not need to appear in the same order within the
 populations.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>other</CODE> - The object to compare against.
<DT><B>Returns:</B><DD>true if the objects are the same, false otherwise.</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=3 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/Genotype.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../org/jgap/Gene.html" title="class in org.jgap"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../org/jgap/IdFactory.html" title="class in org.jgap"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Genotype.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>

</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
