<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.4.2_09) on Sun Sep 04 19:18:59 CDT 2005 -->
<TITLE>
Uses of Package org.jgap (ANJI API)
</TITLE>


<LINK REL ="stylesheet" TYPE="text/css" HREF="../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="Uses of Package org.jgap (ANJI API)";
}
</SCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=3 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Class</FONT>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="package-use.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>

</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<CENTER>
<H2>
<B>Uses of Package<br>org.jgap</B></H2>
</CENTER>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Packages that use <A HREF="../../org/jgap/package-summary.html">org.jgap</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#com.anji.gomoku"><B>com.anji.gomoku</B></A></TD>
<TD><code>com.anji.gomoku</code> contains classes to handle Go-Moku (see <a href="http://www.gamerz.net/pbmserv/gomoku.html">Gamerz.net</a> for rules).&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#com.anji.imaging"><B>com.anji.imaging</B></A></TD>
<TD><code>com.anji.imaging</code> contains utility classes to handle image processing.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#com.anji.integration"><B>com.anji.integration</B></A></TD>
<TD><code>com.anji.integration</code> contains classes that provide the glue between the various components of the system, including artificial neural net implementations, <a href="http://jgap.sourceforge.net/">JGAP</a>, <a href="http://www.cs.utexas.edu/users/kstanley/neat.html">NEAT</a>, Log4J, and persistence and presentation layers.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#com.anji.neat"><B>com.anji.neat</B></A></TD>
<TD><code>com.anji.neat</code> contains classes implementing <a href="http://www.cs.utexas.edu/users/kstanley/neat.html">NEAT</a> framework, including genes, mutation operators, and fitness function used for XOR.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#com.anji.persistence"><B>com.anji.persistence</B></A></TD>
<TD><code>com.anji.persistence</code> contains classes implementing a simple persistence interface.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#com.anji.polebalance"><B>com.anji.polebalance</B></A></TD>
<TD><code>com.anji.polebalance</code> contains classes to implement pole balancing.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#com.anji.roshambo"><B>com.anji.roshambo</B></A></TD>
<TD><code>com.anji.roshambo</code> contains classes to implement the game of 
<a href="http://en.wikipedia.org/wiki/Roshambo">roshambo</a>, also commonly known as 
rock-scissors-paper.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#com.anji.run"><B>com.anji.run</B></A></TD>
<TD><code>com.anji.run</code> contains classes concerned with an evolutionary run.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#com.anji.tournament"><B>com.anji.tournament</B></A></TD>
<TD><code>com.anji.tournament</code> contains classes to handle tournaments and games.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#com.anji.ttt"><B>com.anji.ttt</B></A></TD>
<TD><code>com.anji.ttt</code> contains classes to handle Tic-Tac-Toe (see <a href="http://yahooligans.yahoo.com/content/games/rules/tictactoe/tichistory.html">Yahooligans</a> for rules).&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#com.anji.util"><B>com.anji.util</B></A></TD>
<TD><code>com.anji.util</code> contains utility classes used in the ANJI framework.&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#examples"><B>examples</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.jgap"><B>org.jgap</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><A HREF="#org.jgap.impl"><B>org.jgap.impl</B></A></TD>
<TD>&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="com.anji.gomoku"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Classes in <A HREF="../../org/jgap/package-summary.html">org.jgap</A> used by <A HREF="../../com/anji/gomoku/package-summary.html">com.anji.gomoku</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Chromosome.html#com.anji.gomoku"><B>Chromosome</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Chromosomes represent potential solutions and consist of a fixed-length collection of genes.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="com.anji.imaging"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Classes in <A HREF="../../org/jgap/package-summary.html">org.jgap</A> used by <A HREF="../../com/anji/imaging/package-summary.html">com.anji.imaging</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/BulkFitnessFunction.html#com.anji.imaging"><B>BulkFitnessFunction</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Bulk fitness functions are used to determine how optimal a group of solutions are relative to
 each other.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Chromosome.html#com.anji.imaging"><B>Chromosome</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Chromosomes represent potential solutions and consist of a fixed-length collection of genes.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="com.anji.integration"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Classes in <A HREF="../../org/jgap/package-summary.html">org.jgap</A> used by <A HREF="../../com/anji/integration/package-summary.html">com.anji.integration</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Allele.html#com.anji.integration"><B>Allele</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Allele contains gene data that can be different for multiple chromosomes with the same gene.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/BulkFitnessFunction.html#com.anji.integration"><B>BulkFitnessFunction</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Bulk fitness functions are used to determine how optimal a group of solutions are relative to
 each other.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Chromosome.html#com.anji.integration"><B>Chromosome</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Chromosomes represent potential solutions and consist of a fixed-length collection of genes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Configuration.html#com.anji.integration"><B>Configuration</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The Configuration class represents the current configuration of plugins and flags necessary
 to execute the genetic algorithm (such as fitness function, natural selector, genetic
 operators, and so on).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Genotype.html#com.anji.integration"><B>Genotype</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Genotypes are fixed-length populations of chromosomes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/NaturalSelector.html#com.anji.integration"><B>NaturalSelector</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Natural selectors are responsible for actually selecting a specified number
 of Chromosome specimens from a population, using the fitness values as a
 guide.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="com.anji.neat"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Classes in <A HREF="../../org/jgap/package-summary.html">org.jgap</A> used by <A HREF="../../com/anji/neat/package-summary.html">com.anji.neat</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Allele.html#com.anji.neat"><B>Allele</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Allele contains gene data that can be different for multiple chromosomes with the same gene.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/BulkFitnessFunction.html#com.anji.neat"><B>BulkFitnessFunction</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Bulk fitness functions are used to determine how optimal a group of solutions are relative to
 each other.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Chromosome.html#com.anji.neat"><B>Chromosome</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Chromosomes represent potential solutions and consist of a fixed-length collection of genes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/ChromosomeMaterial.html#com.anji.neat"><B>ChromosomeMaterial</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This is the guts of the original Chromosome object, pulled out so the genes can be modified
 by genetic operators before creating the Chromosome object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Configuration.html#com.anji.neat"><B>Configuration</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The Configuration class represents the current configuration of plugins and flags necessary
 to execute the genetic algorithm (such as fitness function, natural selector, genetic
 operators, and so on).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/CrossoverReproductionOperator.html#com.anji.neat"><B>CrossoverReproductionOperator</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Abstract crossover reporduction operator handles iteration over population to determine pairs
 of parent chromosomes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Gene.html#com.anji.neat"><B>Gene</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Genes represent the discrete components of a potential solution (the Chromosome).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/InvalidConfigurationException.html#com.anji.neat"><B>InvalidConfigurationException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This exception is typically thrown when an invalid value has been
 passed to a Configuration object, an attempt is made to lock a Configuration
 object before all required settings have been provided, or an attempt is
 made to alter a setting in a Configuration object after it has been
 successfully locked.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/MutationOperator.html#com.anji.neat"><B>MutationOperator</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Abstract class for mutation operators.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/ReproductionOperator.html#com.anji.neat"><B>ReproductionOperator</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Abstract class for reproduction operators.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="com.anji.persistence"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Classes in <A HREF="../../org/jgap/package-summary.html">org.jgap</A> used by <A HREF="../../com/anji/persistence/package-summary.html">com.anji.persistence</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Chromosome.html#com.anji.persistence"><B>Chromosome</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Chromosomes represent potential solutions and consist of a fixed-length collection of genes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Configuration.html#com.anji.persistence"><B>Configuration</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The Configuration class represents the current configuration of plugins and flags necessary
 to execute the genetic algorithm (such as fitness function, natural selector, genetic
 operators, and so on).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Genotype.html#com.anji.persistence"><B>Genotype</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Genotypes are fixed-length populations of chromosomes.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="com.anji.polebalance"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Classes in <A HREF="../../org/jgap/package-summary.html">org.jgap</A> used by <A HREF="../../com/anji/polebalance/package-summary.html">com.anji.polebalance</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/BulkFitnessFunction.html#com.anji.polebalance"><B>BulkFitnessFunction</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Bulk fitness functions are used to determine how optimal a group of solutions are relative to
 each other.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Chromosome.html#com.anji.polebalance"><B>Chromosome</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Chromosomes represent potential solutions and consist of a fixed-length collection of genes.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="com.anji.roshambo"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Classes in <A HREF="../../org/jgap/package-summary.html">org.jgap</A> used by <A HREF="../../com/anji/roshambo/package-summary.html">com.anji.roshambo</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Chromosome.html#com.anji.roshambo"><B>Chromosome</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Chromosomes represent potential solutions and consist of a fixed-length collection of genes.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="com.anji.run"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Classes in <A HREF="../../org/jgap/package-summary.html">org.jgap</A> used by <A HREF="../../com/anji/run/package-summary.html">com.anji.run</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Genotype.html#com.anji.run"><B>Genotype</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Genotypes are fixed-length populations of chromosomes.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="com.anji.tournament"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Classes in <A HREF="../../org/jgap/package-summary.html">org.jgap</A> used by <A HREF="../../com/anji/tournament/package-summary.html">com.anji.tournament</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/BulkFitnessFunction.html#com.anji.tournament"><B>BulkFitnessFunction</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Bulk fitness functions are used to determine how optimal a group of solutions are relative to
 each other.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Chromosome.html#com.anji.tournament"><B>Chromosome</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Chromosomes represent potential solutions and consist of a fixed-length collection of genes.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="com.anji.ttt"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Classes in <A HREF="../../org/jgap/package-summary.html">org.jgap</A> used by <A HREF="../../com/anji/ttt/package-summary.html">com.anji.ttt</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Chromosome.html#com.anji.ttt"><B>Chromosome</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Chromosomes represent potential solutions and consist of a fixed-length collection of genes.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="com.anji.util"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Classes in <A HREF="../../org/jgap/package-summary.html">org.jgap</A> used by <A HREF="../../com/anji/util/package-summary.html">com.anji.util</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/BulkFitnessFunction.html#com.anji.util"><B>BulkFitnessFunction</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Bulk fitness functions are used to determine how optimal a group of solutions are relative to
 each other.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Configuration.html#com.anji.util"><B>Configuration</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The Configuration class represents the current configuration of plugins and flags necessary
 to execute the genetic algorithm (such as fitness function, natural selector, genetic
 operators, and so on).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/ReproductionOperator.html#com.anji.util"><B>ReproductionOperator</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Abstract class for reproduction operators.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="examples"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Classes in <A HREF="../../org/jgap/package-summary.html">org.jgap</A> used by <A HREF="../../examples/package-summary.html">examples</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Chromosome.html#examples"><B>Chromosome</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Chromosomes represent potential solutions and consist of a fixed-length collection of genes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/FitnessFunction.html#examples"><B>FitnessFunction</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<I>use <code>BulkFitnessFunction</code> instead</I></TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.jgap"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Classes in <A HREF="../../org/jgap/package-summary.html">org.jgap</A> used by <A HREF="../../org/jgap/package-summary.html">org.jgap</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Allele.html#org.jgap"><B>Allele</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Allele contains gene data that can be different for multiple chromosomes with the same gene.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/BulkFitnessFunction.html#org.jgap"><B>BulkFitnessFunction</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Bulk fitness functions are used to determine how optimal a group of solutions are relative to
 each other.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Chromosome.html#org.jgap"><B>Chromosome</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Chromosomes represent potential solutions and consist of a fixed-length collection of genes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/ChromosomeMaterial.html#org.jgap"><B>ChromosomeMaterial</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This is the guts of the original Chromosome object, pulled out so the genes can be modified
 by genetic operators before creating the Chromosome object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Configuration.html#org.jgap"><B>Configuration</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The Configuration class represents the current configuration of plugins and flags necessary
 to execute the genetic algorithm (such as fitness function, natural selector, genetic
 operators, and so on).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/FitnessFunction.html#org.jgap"><B>FitnessFunction</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<B>Deprecated.</B>&nbsp;<I>use <code>BulkFitnessFunction</code> instead</I></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Gene.html#org.jgap"><B>Gene</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Genes represent the discrete components of a potential solution (the Chromosome).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Genotype.html#org.jgap"><B>Genotype</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Genotypes are fixed-length populations of chromosomes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/IdFactory.html#org.jgap"><B>IdFactory</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Factory for generating unique IDs across multiple runs.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/InvalidConfigurationException.html#org.jgap"><B>InvalidConfigurationException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This exception is typically thrown when an invalid value has been
 passed to a Configuration object, an attempt is made to lock a Configuration
 object before all required settings have been provided, or an attempt is
 made to alter a setting in a Configuration object after it has been
 successfully locked.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/MutationOperator.html#org.jgap"><B>MutationOperator</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Abstract class for mutation operators.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/NaturalSelector.html#org.jgap"><B>NaturalSelector</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Natural selectors are responsible for actually selecting a specified number
 of Chromosome specimens from a population, using the fitness values as a
 guide.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/ReproductionOperator.html#org.jgap"><B>ReproductionOperator</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Abstract class for reproduction operators.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/SpeciationParms.html#org.jgap"><B>SpeciationParms</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Specie.html#org.jgap"><B>Specie</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Species are reproductively isolated segments of a population.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<A NAME="org.jgap.impl"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
Classes in <A HREF="../../org/jgap/package-summary.html">org.jgap</A> used by <A HREF="../../org/jgap/impl/package-summary.html">org.jgap.impl</A></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Allele.html#org.jgap.impl"><B>Allele</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Allele contains gene data that can be different for multiple chromosomes with the same gene.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Chromosome.html#org.jgap.impl"><B>Chromosome</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Chromosomes represent potential solutions and consist of a fixed-length collection of genes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/Configuration.html#org.jgap.impl"><B>Configuration</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The Configuration class represents the current configuration of plugins and flags necessary
 to execute the genetic algorithm (such as fitness function, natural selector, genetic
 operators, and so on).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/NaturalSelector.html#org.jgap.impl"><B>NaturalSelector</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Natural selectors are responsible for actually selecting a specified number
 of Chromosome specimens from a population, using the fitness values as a
 guide.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/ReproductionOperator.html#org.jgap.impl"><B>ReproductionOperator</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Abstract class for reproduction operators.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><B><A HREF="../../org/jgap/class-use/UnsupportedRepresentationException.html#org.jgap.impl"><B>UnsupportedRepresentationException</B></A></B>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This exception is typically thrown when the
 setValueFromPersistentRepresentation() method of a Gene class is unable
 to process the string representation it has been given, either because that
 representation is not supported by that Gene implementation or because
 the representation is corrupt.</TD>
</TR>
</TABLE>
&nbsp;
<P>
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=3 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Class</FONT>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Use</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV&nbsp;
&nbsp;NEXT</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="package-use.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>

</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
